server/extension/sql/21_bulk_geocode_street.sql

```
@@ -0,0 +1,75 @@
+      google_plan = plpy.prepare("SELECT * FROM cdb_dataservices_server._cdb_bulk_google_geocode_street_point($1, $2, $3); ", ["text", "text", "jsonb"])
+      result = plpy.execute(google_plan, [username, orgname, searches])
+      return result
+    else:
+      raise Exception('Requested geocoder is not available')
+
+$$ LANGUAGE plpythonu STABLE PARALLEL RESTRICTED;
+
+CREATE OR REPLACE FUNCTION cdb_dataservices_server._cdb_bulk_google_geocode_street_point(username TEXT, orgname TEXT, searches jsonb)
+RETURNS SETOF cdb_dataservices_server.geocoding AS $$
+  from cartodb_services.tools import LegacyServiceManager,QuotaExceededException,Logger
+  from cartodb_services.google import GoogleMapsGeocoder
+
+  plpy.execute("SELECT cdb_dataservices_server._get_logger_config()")
+  logger_config = GD["logger_config"]
+
+  logger = Logger(logger_config)
+  service_manager = LegacyServiceManager('geocoder', username, orgname, GD)
+
+  try:
+    service_manager.assert_within_limits(quota=False)
+    geocoder = GoogleMapsGeocoder(service_manager.config.google_client_id, service_manager.config.google_api_key, service_manager.logger)
+    geocode_results = geocoder.bulk_geocode(searches=searches)
+    if geocode_results:
+      results = []
+      for result in geocode_results:
+        if result[1]:
+          plan = plpy.prepare("SELECT ST_SetSRID(ST_MakePoint($1, $2), 4326) as the_geom; ", ["double precision", "double precision"])
+          point = plpy.execute(plan, result[1], 1)[0]
+          results.append([result[0], point['the_geom'], None])
+        else:
+          results.append([result[0], None, None])
+      service_manager.quota_service.increment_success_service_use(len(results))
+      return results
+    else:
+      service_manager.quota_service.increment_empty_service_use(len(searches))
+      return []
+  except QuotaExceededException as qe:
+    service_manager.quota_service.increment_failed_service_use(len(searches))
```

What happens if it just fails for the first element in the search? Do we consider the full batch as failed?

---

For quota, we do. But if you check `bulk_geocode`, it swallows most of the errors, so virtually no individual error should make the batch fail.